### PR TITLE
[ML][FEATURE] Use memory limit to assign data frame analytics jobs

### PR DESCRIPTION
**Pull request by droberts195**
This change switches data frame analytics jobs from
simple assignment to the first available ML node to
assignment based on the memory requirement of the
data frame analytics job and the available memory on
each ML nodes.

The memory requirement of a data frame analytics job
is now defined to be its model memory limit plus 20MB
for process overhead.

Available memory on each ML node takes into account
running anomaly detector jobs as well as data frame
analytics jobs.  To facilitate this the shared logic has
been moved from TransportOpenJobAction to a new
class, JobNodeSelector.

### CODE REVIEW

**Review by dimitris-athanasiou — COMMENTED** [review]
Looks great! Just left a couple of comments.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStartDataFrameAnalyticsAction.java [hunk 1]
```diff
@@ -280,10 +300,45 @@ protected AllocatedPersistentTask createTask(
...
+                }
+            }
+
+            JobNodeSelector jobNodeSelector = new JobNodeSelector(clusterState, id, MlTasks.DATA_FRAME_ANALYTICS_TASK_NAME, memoryTracker,
+                node -> nodeFilter(node, id));
+            // Pass an effectively infinite value for max concurrent opening jobs, because data frame analytics jobs do
+            // not have an "opening" state so would never be rejected for causing too many jobs in the "opening" state
+            PersistentTasksCustomMetaData.Assignment assignment = jobNodeSelector.selectNode(
+                maxOpenJobs, Integer.MAX_VALUE, maxMachineMemoryPercent, isMemoryTrackerRecentlyRefreshed);
+            if (assignment.getExecutorNode() == null) {
```
[the remark this answers is not in the document]
Comment (droberts195) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStartDataFrameAnalyticsAction.java]: Yes, good point.  And while doing this I realised I left `AWAITING_LAZY_ASSIGNMENT` out of the analytics predicate, so I've added it there too.

**Review by dimitris-athanasiou — APPROVED** [review]
LGTM

**Comment by droberts195** [comment]
Now that elastic/ml-cpp#454 is merged, PR CI for this PR will not pass until #40927 is merged.

**Pull request merged**